src/Infrastructure/LeanCode.Firebase.FCM/NotificationDataConverter.cs

```diff
@@ -0,0 +1,58 @@
+using System.Collections.Immutable;
+using System.Reflection;
+
+namespace LeanCode.Firebase.FCM;
+
+public sealed class NotificationDataConverter
+{
+    private const string TypeField = "Type";
+
+    private ImmutableDictionary<Type, Func<object, string>> formatters = ImmutableDictionary<
+        Type,
+        Func<object, string>
+    >.Empty;
+
+    public void FormatUsing<T>(Func<T, string> formatter)
```

Comment:
I was hoping we would require all formatters to be provided at construction time and `NotificationDataConverter` could then be immutable. Can we do this, using a builder pattern or something?

Reply:
+1 from me

Reply:
A great overkill but can be done.

Reply:
Plus, if it's assumed to be a singleton you could then use `FrozenDictionary` for storage.